### PR TITLE
V0.45.13 proposal

### PR DESCRIPTION
**Pull request by nghuyenthevinh2000**
## Summary of changes

No chain migration code is needed

There is only REST endpoints migration: https://github.com/cosmos/cosmos-sdk/blob/v0.45.13/docs/migrations/rest.md. This upgrade will break all current front - end.

## Report of required housekeeping

- [ ] Github issue OR spec proposal link
- [ ] Wrote tests
- [ ] Updated API documentation (client/lcd/swagger-ui/swagger.yaml)
- [ ] Added a relevant changelog entry: clog add [section] [stanza] [message]

----

## (FOR ADMIN) Before merging

- [ ] Added appropriate labels to PR
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" (coding standards)
- [ ] Confirm added tests are consistent with the intended behavior of changes
- [ ] Ensure all tests pass

### CODE REVIEW

**Comment by fragwuerdig** [comment]
@inon-man @nghuyenthevinh2000: What implications has this for Mantlemint and FCD?

**Review by ZaradarBH — APPROVED** [review]
LGTM buddy. Great work! :)

**Review by ZaradarBH — APPROVED** [review]

**Comment by nghuyenthevinh2000** [comment]
@fragwuerdig 
can I have your approval on this one?

**Comment by fragwuerdig** [comment]
> figured out the right way to modify the gas test. If you can check the gas test, you will see that it is very minimal

Do you have an idea what we need to modify in FCD so that we do not break the Infra?

File: x/wasm/types/gas_meter_test.go
Comment (fragwuerdig) [inline x/wasm/types/gas_meter_test.go:1]: Can you short explain why you needed to do that?

**Comment by faddat** [comment]
Hey I heard there was need of a review here. So normally I would totally lint this, before merging it.

If you're cool with waiting until tomorrow, then I'm happy to review. What I need to know in order to review this effectively, is the style and goal of this particular pull request.

@nghuyenthevinh2000 can you explain to me, what the goals are here, and how you intend to deal with WASM in this PR?

File: x/wasm/types/gas_meter_test.go
Comment (faddat) [reply x/wasm/types/gas_meter_test.go:1]: If I had to guess, this was necessary because SDK 45 changes the way the SDK consumes gas.

**Comment by faddat** [comment]
I think that we should be using a later version of TMDb, that uses grocksdb instead of gorocksdb.

**Comment by faddat** [comment]
Tendermint 34.24 has a CPU griefing issue, and the minimum version that I consider safe is 

GitHub.com/informalsystems/tendermint v0.34.26

**Comment by faddat** [comment]
Since we are updating the cosmos SDK, the minimum version that we should be updating to is v0.45.14

File: go.mod
Comment (faddat) [inline go.mod:133]: Should be able to just specify the version that you want above, instead of replacing here.

File: go.mod
Comment (faddat) [inline go.mod:129]: Since you're targeting only one version, you should remove the version number from one of these replacements, and then specify only v1.33.0

If you actually need several different versions then you should refactor upstream

File: go.mod
Comment (faddat) [inline go.mod:130]: 0.45.14

And .15 if you'd like to use comet and comet is probably a good idea.

File: go.mod
Comment (faddat) [inline go.mod:137]: You should be able to specify the version that you desire above, instead of replacing

File: go.mod
Comment (faddat) [inline go.mod:138]: You should be able to specify the version that you would like to use above, instead of replacing

File: x/wasm/keeper/submsg_test.go
Comment (faddat) [inline x/wasm/keeper/submsg_test.go:298]: I'm going to go ahead and assume that this is here and changed because SDK 45 consumes gas differently from 44. If that's the case and you're sure these figures are correct, then everything is fine here

File: go.mod
Comment (faddat) [inline go.mod:140]: This version of tendermint contains a CPU griefing bug.

File: go.mod
Comment (faddat) [inline go.mod:108]: If you look at go.mod in WASMD, main branch, you're going to find that there have been issues with go level db using inconsistent versions and that the way to fix this is using a replacement.

File: go.mod
Comment (faddat) [inline go.mod:43]: Good

**Comment by faddat** [comment]
I merged this branch into the linted one. This makes it easier to find issues.  Remember to turn on golangci-lint in your editor.

**Comment by faddat** [comment]
Basically I don't approve this PR, until:

* it has been linted
* it no longer uses vulnerable software
* the replace section of go.mod is limited more effectively

* #178 should be merged here before this is considered for merge to main.

**Comment by ZaradarBH** [comment]
> Hey I heard there was need of a review here. So normally I would totally lint this, before merging it.
> 
> If you're cool with waiting until tomorrow, then I'm happy to review. What I need to know in order to review this effectively, is the style and goal of this particular pull request.
> 
> @nghuyenthevinh2000 can you explain to me, what the goals are here, and how you intend to deal with WASM in this PR?

Thank you for providing feedback. The more the merrier, just lets keep it civil and remember we all have the same goal here <3

We intend to leave the TFL WASM customization in place for the 2.0 upgrade, since we want to make sure we can have complete focus on the WASM migration process that L2 apps will have to go thru. Thus we will revert to the canonical Cosmos SDK WASM module in 2.1.

The 2.0 / 2.1 upgrades will break the FCD but as such I believe that it will mainly be the protobuf message contracts so it will be fairly simple to upgrade the FCD TypeScript dependencies and do whatever tweaking is needed once we have the new message contract dependency in place (terra.proto / terra.js). Terra-station, desktop, etc will also be affected by the changes to the message contracts. But the 0.45.x upgrade is largely just extensions to existing msg schemas so it should be mainly about rewiring code, thus it should not take to long to fix.

Regarding this comment: 

"Tendermint 34.24 has a CPU griefing issue, and the minimum version that I consider safe is

GitHub.com/informalsystems/tendermint v0.34.26"

As I can see it TM 34.26 is really just "pre-stage-CometBFT", since its released by informalsystems and not the official tendermint release? Either way, I tried bumping 0.45.13 to use 34.26 but it seemed to introduce some breakage due to some issues related to github.com/gogo/protobuf v1.3.3 and ultimately a missing replace statement for informals TM version. However I cannot find any reference to the "CPU griefing issue" in their CHANGELOG, do you have a link for some info on that particular issue?

It should be safe to bump to Cosmos SDK v0.45.14, which has a DDoS patch of some sort. However I think we can do that safely in the context of your clean up PR, Granted that you downgrade to Go 1.18, as we don't have the resources atm to help validators update their go installations and deal with potential "unknown, unknowns", which we most likely will have to do if we make this a hard requirement.

File: go.mod
Comment (ZaradarBH) [reply go.mod:128]: These type of replacements exist to force dependencies-of-indirect-dependencies to use specific package versions to deal with CVEs that I monitor across our different repos via https://app.snyk.io/org/l1-task-force/project/18142370-e75d-47bc-878d-cc9c3439b99e . I did an initial scan on the codebase last summer and wrote this postmortem @ https://github.com/terra-rebels/postmortems/blob/main/06272022-SecOps-Review-Postmortem.md , unfortunately this is a never ending "hamster wheel" of CVE patching so I think we will end up having a lot more of these types of "assembly redirects" if we want to eliminate all of the potential high and medium severity attack vectors

File: go.mod
Comment (ZaradarBH) [reply go.mod:129]: Same as above. I am retargeting two different package version requests from two indirect dependencies to fix CVE issues

File: go.mod
Comment (ZaradarBH) [reply go.mod:130]: We will be looking into CometBFT after 2.1. To ensure we do things in small increments

File: go.mod
Comment (ZaradarBH) [reply go.mod:137]: Same as other CVE comments

File: go.mod
Comment (ZaradarBH) [reply go.mod:138]: Same as other CVE comments

File: go.mod
Comment (ZaradarBH) [reply go.mod:140]: Is this the CPU griefing issue your refering too? https://github.com/informalsystems/tendermint/issues/4

Comment (ZaradarBH) [reply go.mod:140]: If so I think it would make more sense to stay with 0.34.24 since the CPU griefing issue seems to be fixed in https://github.com/cometbft/cometbft/pull/245 so we should get this automagically when we release 2.2 and migrate to CometBFT

File: go.mod
Comment (ZaradarBH) [reply go.mod:141]: We have to replace to ensure that all requests for tendermint packages hits our modified version containing the IsOracleTx logic

File: x/wasm/types/gas_meter_test.go
Comment (nghuyenthevinh2000) [reply x/wasm/types/gas_meter_test.go:1]: yep, SDK changes lead to wasm gas change

Comment (ZaradarBH) [reply x/wasm/types/gas_meter_test.go:1]: @nghuyenthevinh2000 updated it to align with the required settings in SDK 45

**Pull request merged**
